src/queue.js

```diff
@@ -0,0 +1,98 @@
+let instance = null
+
+class Queue {
+
+  constructor() {
+    if(!instance){
+      instance = this // eslint-disable-line consistent-this
+      this.setup()
+    }
+    return instance
+  }
+
+  setup() {
+    this.tasks = []
+    this.pendingCounter = {}
+    this.ranCounter = {}
+    this.running = false
+  }
+
+  valid(task) {
+    if(
+      typeof task.run !== 'function' ||
+      typeof task.error !== 'function' ||
+      typeof task.done !== 'function'
+    ) throw new Error('Task has an invalid object signature, methods required')
+    return true
+  }
+
+  incrementQueue(queue) {
+    if(!this.pendingCounter[queue]) {
+      this.pendingCounter[queue] = 0
+      this.ranCounter[queue] = 0
+    }
+    this.pendingCounter[queue] = this.pendingCounter[queue] + 1
+  }
+
+  incrementRuns(queue) {
+    if(this.ranCounter[queue] === 0)
+    this.ranCounter[queue] = this.ranCounter[queue] + 1
+  }
+
+  decrementQueue(queue) {
+    this.pendingCounter[queue] = this.pendingCounter[queue] - 1
+    if(this.pendingCounter[queue] === 0){
+      this.ranCounter[queue] = 0
+    }
+  }
+
+  start() {
+    if(this.tasks.length > 0)
+      this.runNext()
+  }
+
+  pushTo(queue, task) {
+    this.push(task, queue)
+  }
+
+  push(task, queue = 'generic') {
+    if(!this.valid(task)) return
```

Comment:
That's a good example of `Queue.valid` throwing when it's expected to return `false`. See comments above.